ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsRange.scala

```diff
@@ -61,4 +66,22 @@ object EventsRange {
       .as(get[Long](1).singleOpt)(connection)
       .getOrElse(EmptyLedgerEventSeqId)
   }
+
+  private[events] def readPage[A](
+      fasterRead: Long => SimpleSql[Row], // takes guessedPageEnd
+      saferRead: Int => SimpleSql[Row], // takes minPageSize
+      row: RowParser[A],
+      range: EventsRange[Long],
+      pageSize: Int): SqlSequence[Vector[A]] = {
+    val minPageSize = 10 min pageSize max (pageSize / 10)
+    val guessedPageEnd = range.endInclusive min (range.startExclusive + pageSize)
+    SqlSequence
+      .vector(fasterRead(guessedPageEnd) withFetchSize Some(pageSize), row)
+      .flatMap { arithPage =>
+        if (guessedPageEnd == range.endInclusive || arithPage.sizeIs >= minPageSize)
+          SqlSequence point arithPage
+        else
+          SqlSequence.vector(saferRead(minPageSize) withFetchSize Some(minPageSize), row)
+      }
+  }
```

Comment:
To restate the general strategy in prose, to make it easier to think about edge cases:

1. We pick a min page size, which is pageSize/10 but at least 10 (unless the pageSize < 10, but let's set that aside).
2. First, query row IDs [startExclusive, startExclusive + pageSize), unless the latter exceeds endInclusive.
    a. If that yields "min page size" elements, then return that as a page. So, consider pageSize = 1000 with a SQL predicate of frequency about 10%. If we find >=100 elements in the initial query, we just return them, but if we find <=99 elements, we query a second time.
    b. Otherwise, do a fallback query. This is exactly like the query that preceded this PR, except we use `limit minPageSize` instead of `limit pageSize`. So, in our example, we would look only for the first 100 elements that match, rather than the first 1000.

Reply:
Probably worth adding this as a comment somewhere, so that it's always where the reader needs it the most. :wink:

Reply:
I'd rather not, because I'm basically asking for it to be outdated within a week. This description is meant more for you to disagree with it so it can become more quickly outdated :)

Reply:
why not just use `pageSize` instead of `minPageSize` in the fallback query? If we are doing the slow query, just return the `pageSize` number of elements.

Reply:
@leo-da The reason is that the predicate is linear, and in the fallback, our search space is `limit / frequency`. In other words, we reduce the search space by 90%, which can be a very large reduction for low-frequency queries.

This is meant to be in harmony with how we determine whether we need to fallback in the first place: if it was unacceptable to only return 99 elements, but would have been acceptable to return 100 elements if the first query had found them, we should only spend time looking for 100 elements.

The Big Idea of doing the arithmetic query is to limit the search space. The goal of the fallback is to avoid problematic cases with the first query, such as returning only one element per page over the ledger API, which would be very bad performance-wise.

Reply:
A party with sparing events will pretty much always trigger the slow query, right?

Reply:
so, limiting the  search space you are optimizing the query, decreasing the latency. However you will have to do more of those fast queries in the case of the sparse party events.

Reply:
@gerolf-da Yes, according to how evenly distributed its events are. (You can think of infrequent template IDs similarly.)

One possible extension might be to feedback that the `limit` query was used on the first page and use it for later queries. That might have other edge cases I'm not thinking of, though.

Some of this can be dealt with by adjusting how minPageSize is chosen. Originally, it was always 1, so the limit query was used only in the unworkable case where the fast query returned no results at all; however, a stream of transaction response messages each containing only 1 event seemed...not great.

Reply:
@leo-da Yeah, trade-offs abound here. I would like to skip the fast query if I was sure that it wouldn't yield enough results, but I'm unsure what the best way to perform that heuristic would be.

Reply:
Slight alteration in 78f8f70af8: if minPageSize is 100, and we found 80 in the first query, we only search for 20 in the second query. I've incidentally eliminated all the duplicate queries I had introduced.

(As you surely agree, reasoning about whether direct SQL interpolation is safe is much better than being able to append fragments of literal SQL and positional arguments.)